### PR TITLE
[sglang] feat: Support async multi-turn rollout with simulation feedback in sglang

### PR DESCRIPTION
**Pull request by kinza99**
### Checklist Before Starting

- [x] Searched for similar PR(s).
- [x] Checked PR Title format
  - In format of: [modules] type: Title
  - modules are in `fsdp, megatron, sglang, vllm, rollout, trainer, ci, training_utils, recipe, hardware, deployment, ray, worker, single_controller, misc, perf, model, algo, env, tool, ckpt, doc, data`
  - type is in `feat, fix, refactor, chore, test`
  - can involve multiple modules, seperated by `,` or space, like `[megatron, fsdp, doc] feat: xxx`

### What does this PR do?

> Implements multi-turn interaction system for reinforcement learning training, enabling dynamic conversational feedback and iterative problem-solving scenarios with extensible agent-based architecture.

### Test

> Comprehensive testing has been performed including:
> - **Unit Tests**: Complete test suite in `tests/interactions/test_gsm8k_interaction.py` with 15+ test cases covering:
>   - Initialization and configuration validation
>   - Interaction lifecycle (start → generate → calculate → finalize)
>   - Correct/incorrect answer handling with GSM8K scoring integration
>   - Edge cases (empty messages, malformed content, concurrent sessions)
>   - Resource cleanup and error handling
> - **Integration Tests**: SGLang rollout integration in `tests/workers/rollout/test_sglang_async_rollout_w_interaction.py`:
>   - Multi-GPU distributed testing (requires 2+ GPUs)
>   - FSDP model sharding with SGLang inference engine
>   - End-to-end interaction workflow with `interaction_kwargs` parameter passing
>   - Comparison with HuggingFace baseline for output validation
> - **Real-world Validation**: Training script testing with Qwen2.5-0.5B model demonstrating:
>   - GRPO algorithm integration with interaction-based rewards
>   - Multi-turn conversation handling in production training loops

### High-Level Design

> **Multi-Turn Interaction System Architecture:**
> 
> The system introduces a flexible, async-based interaction framework designed for RL training scenarios with the following key components:
> 
> 1. **BaseInteraction Class**: Core abstraction layer providing async interface for interaction agents
> 2. **Instance Management**: Stateful session management with unique instance IDs for concurrent interactions
> 3. **SGLang Integration**: Seamless integration with SGLang rollout system for multi-turn conversations
> 4. **Configuration-Driven Loading**: Dynamic agent loading via YAML configuration files
> 5. **Reward Integration**: Turn-level scoring mechanism integrated with VERL's reward system

### Specific Changes

> **Core Implementation:**
> - Added `BaseInteraction` abstract class with async interface in `verl/interactions/base.py`
> - Implemented `Gsm8kInteraction` concrete class for math problem solving scenarios in `verl/interactions/gsm8k_interaction.py`
> - Added instance-based state management for concurrent interaction sessions via `_instance_dict`
> - Created turn-level scoring mechanism using `verl.utils.reward_score.gsm8k` with flexible answer extraction
> 
> **SGLang Rollout Integration (`verl/workers/rollout/sglang_rollout/sglang_rollout.py`):**
> - Extended `_async_rollout_a_request` method with `AsyncRolloutRequestStateEnum.INTERACTING` state
> - Added interaction initialization in `_handle_pending_state` method (line 878-880)
> - Implemented interaction response generation with reward accumulation (line 823-835)
> - Dynamic interaction loading via `_intitalize_interaction` method with importlib-based class resolution
> - Configuration-driven setup supporting `interaction_config_path` parameter
> 
> **Testing Infrastructure:**
> - Comprehensive unit test suite in `tests/interactions/test_gsm8k_interaction.py` with pytest and asyncio support
> - Integration tests in `tests/workers/rollout/test_sglang_async_rollout_w_interaction.py` with distributed GPU testing
> - Mock-based testing using `unittest.mock.patch` for GSM8K scoring validation
> 
> **Configuration & Examples:**
> - GSM8K interaction configuration in `examples/sglang_multiturn/config/interaction_config/gsm8k_interaction_config.yaml`
> - Training script: `run_qwen2.5-0.5b_gsm8k_multiturn_w_interaction.sh` with interaction_config_path parameter
> - Support for `max_user_turns` and `max_assistant_turns` configuration

### API

> **BaseInteraction Interface:**
> 
> ```python
> from verl.interactions.base import BaseInteraction
> from typing import Dict, Any, List, Tuple, Optional
> 
> class CustomInteraction(BaseInteraction):
>     def __init__(self, config: Dict[str, Any]):
>         super().__init__(config)
>     
>     async def start_interaction(self, instance_id: Optional[str] = None, **kwargs) -> str:
>         # Initialize interaction session, return instance_id
>         pass
>     
>     async def generate_response(self, instance_id: str, messages: List[Dict[str, Any]], **kwargs) -> Tuple[bool, str, float, Dict[str, Any]]:
>         # Generate response, return (should_terminate, response, score, metadata)
>         pass
>     
>     async def calculate_score(self) -> float:
>         # Calculate turn-level score for RL training
>         pass
>         
>     async def finalize_interaction(self) -> None:
>         # Clean up resources
>         pass
> ```

### Usage Example

> **GSM8K Interaction Configuration:**
> 
> ```yaml
> # gsm8k_interaction_config.yaml
> interaction:
>   - class_name: "verl.interactions.gsm8k_interaction.Gsm8kInteraction"
>     config: {}
> ```
> 
> **Training Script Integration:**
> 
> ```bash
> python3 -m verl.trainer.main_ppo \
>     --config-path="$CONFIG_PATH" \
>     --config-name='gsm8k_multiturn_grpo_w_interaction' \
>     algorithm.adv_estimator=grpo \
>     data.train_batch_size=512 \
>     data.return_raw_chat=True \
>     actor_rollout_ref.rollout.name=sglang \
>     actor_rollout_ref.rollout.multi_turn.interaction_config_path="$PROJECT_DIR/examples/sglang_multiturn/config/interaction_config/gsm8k_interaction_config.yaml" \
>     trainer.total_epochs=15
> ```
> 
> **GSM8K Interaction Implementation:**
> 
> ```python
> class Gsm8kInteraction(BaseInteraction):
>     def __init__(self, config: dict):
>         super().__init__(config)
>         self._instance_dict = {}
> 
>     async def start_interaction(self, instance_id=None, ground_truth=None, **kwargs):
>         if instance_id is None:
>             instance_id = str(uuid4())
>         self._instance_dict[instance_id] = {
>             "response": "",
>             "ground_truth": ground_truth,
>             "reward": 0.0,
>         }
>         return instance_id
> 
>     async def generate_response(self, instance_id, messages, **kwargs):
>         # Extract last user message content
>         content = ""
>         for item in reversed(messages):
>             if item.get("role") == "user":
>                 content = item.get("content", "")
>                 break
> 
>         # Ensure GSM8K format (#### prefix)
>         if content.startswith("#### "):
>             self._instance_dict[instance_id]["response"] = content
>         else:
>             self._instance_dict[instance_id]["response"] = "#### " + content
> 
>         reward = await self.calculate_score(instance_id)
>         if reward == 1.0:
>             return True, "Your response is correct!", 1.0, {}
>         else:
>             return False, "Your response is incorrect! You need to reflect on your answer and try again.", 0.0, {}
> 
>     async def calculate_score(self, instance_id, **kwargs):
>         return gsm8k.compute_score(
>             self._instance_dict[instance_id]["response"],
>             self._instance_dict[instance_id]["ground_truth"],
>             method="flexible", format_score=0.0, score=1.0,
>         )
> ```

### Contributor List

- **He Du** (Author)
- **Xiang Long** (Co-author)
- **Yanbin Jiang** (Discussed integrated with LangGraph scene)
- **Junrong Lin** (Reviewer)
- **Haibin Lin** (Reviewer)
- **Chenyang Zhao** (PM)

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [ ] Add `[BREAKING]` to the PR title `description` if it breaks any API.
- [x] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [x] New CI unit test(s) are added to cover the code path.
- [x] Rely on existing unit tests on CI that covers the code path.

### CODE REVIEW

**Comment by kinza99** [comment]
@zhaochenyang20

**Comment by zhaochenyang20** [comment]
We will take a look these days, stay tuned in 24h
.

**Comment by zhaochenyang20** [comment]
@kinza99 my wechat is 18015766633. Feel free to discuss, thanks!

File: examples/data_preprocess/gsm8k_multiturn_w_tool.py [hunk 1]
```diff
@@ -92,6 +92,10 @@ def process_fn(example, idx):
                             # "release_kwargs": {},
                         },
                     },
+                    "feedback_kwargs": {
```
Comment (SwordFaith) [inline examples/data_preprocess/gsm8k_multiturn_w_tool.py]: It might be better to rename it to "interaction_kwargs".

File: docs/sglang_multiturn/multiturn.rst [hunk 2]
```diff
@@ -40,6 +40,19 @@ Finally, set the ``tools_config_file`` in your rollout config:
...
+
+    feedback:
+      - class_name: ""
+        config: {}
+
+.. code-block:: yaml
+
+    actor_rollout_ref:
+        rollout:
+            feedback_config_file: <path_to_feedback_yaml_file>
```
Comment (SwordFaith) [inline docs/sglang_multiturn/multiturn.rst]: interaction_config_file

File: verl/interactions/base.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        self.config = config
+        self.name: str = config.get("name", "interaction_agent")  # More general agent default role name
+
+    async def start_interaction(self) -> str:  # More clear interaction start method
+        """
+        Initializes a new interaction session and returns its unique ID.
+        Simulates: get id + state init
+        """
+        # ...implement the logic to get ID and initialize state...
+        interaction_id = "some_unique_id"
```
Comment (SwordFaith) [inline verl/interactions/base.py]: It might be better support uuid by default like https://github.com/volcengine/verl/blob/main/verl/tools/base_tool.py#L50.

File: verl/interactions/base.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        return interaction_id
+
+    async def generate_response(self, messages: Any) -> Tuple[bool, str, float, Dict[str, Any]]:  # More clear response generation method
```
Comment (SwordFaith) [inline verl/interactions/base.py]: Sorry for missing instance_id in doc, it would be better keep instance id logic in tool to track session.

File: verl/interactions/gsm8k_interaction.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                break
+
+        if content.startswith("#### "):
+            self._instance_dict[instance_id]["response"] = content
+        else:
+            self._instance_dict[instance_id]["response"] = "#### " + content
+
+        reward = await self.calculate_score(instance_id)
+        if reward == 1.0:
+            feedback = "Your response is correct!"
```
Comment (SwordFaith) [inline verl/interactions/gsm8k_interaction.py]: feedback -> response

File: verl/workers/rollout/sglang_rollout/async_sglang_rollout.py [hunk 6]
```diff
@@ -560,11 +590,35 @@ async def _async_rollout_a_request(self, req: AsyncRolloutRequest, do_sample: bo
...
+                            else:
+                                _req.state = AsyncRolloutRequestStateEnum.COMPLETED
+                                break
                     else:
                         _req.add_assistant_message(self.tokenizer, content, format=self.config.multi_turn.format)
+                        if _req.interaction_kwargs is not None and user_turns < self.config.multi_turn.user_max_turns and current_turns < self.config.multi_turn.max_turns:
+                            _req.state = AsyncRolloutRequestStateEnum.WAITING
+                        else:
+                            break
+            elif _req.state == AsyncRolloutRequestStateEnum.WAITING:
```
Comment (SwordFaith) [inline verl/workers/rollout/sglang_rollout/async_sglang_rollout.py]: WAITING is a bad state name, is there any other concrete state name for interacting with Interaction cls?

File: verl/workers/rollout/sglang_rollout/async_sglang_rollout.py [hunk 7]
```diff
@@ -560,11 +590,35 @@ async def _async_rollout_a_request(self, req: AsyncRolloutRequest, do_sample: bo
...
+                messages = [{"role": x.role, "content": x.content} for x in _req.messages]
+                should_terminate_sequence, content, reward, metrics = await self.interaction.generate_response(_req.request_id, messages, **_req.interaction_kwargs)
+                user_turn_rewards.append(reward)
+                if should_terminate_sequence:
+                    finish_reason_type = FinishReasonTypeEnum.STOP
+                    _req.state = AsyncRolloutRequestStateEnum.COMPLETED
+                    break
+                else:
+                    _req.add_user_message(self.tokenizer, content, format=self.config.multi_turn.format)
+                    if len(_req.input_ids) >= self.config.max_model_len:
```
Comment (SwordFaith) [inline verl/workers/rollout/sglang_rollout/async_sglang_rollout.py]: Need add unit test for new added interaction related logic.

File: verl/workers/actor/dp_actor.py [hunk 8]
```diff
@@ -80,7 +80,7 @@ def _forward_micro_batch(self, micro_batch, temperature, calculate_entropy=False
         """
         response_length = micro_batch["responses"].size(-1)
         multi_modal_inputs = {}
-        if "multi_modal_inputs" in micro_batch:
+        if "multi_modal_inputs" in micro_batch.keys():
```
Comment (ocss884) [inline verl/workers/actor/dp_actor.py]: Just curious why adding lot `.keys()` in this PR? For `... in dict_like_obj` the statements are equivalent

File: verl/interactions/base.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    async def calculate_score(self) -> float:  # More clear score calculation method
+        """
+        Calculates a score for the interaction,
+        potentially considering aspects like partial exposure & in-context task switching.
+        should be invoke at turn-level
+        """
+        # ...implement the logic to calculate turn-level score...
+        score = 0.0
+        return score
```
Comment (ocss884) [inline verl/interactions/base.py:62]: How about `raise NotImplementedError`

File: verl/workers/fsdp_workers.py [hunk 10]
```diff
@@ -15,20 +15,25 @@
...
 import logging
 import os
 import warnings
+from dataclasses import asdict
 from typing import Union
 
 import psutil
 import torch
 import torch.distributed
+import torch.distributed as dist
```
Comment (ocss884) [inline verl/workers/fsdp_workers.py]: Duplicate torch.distributed

File: verl/interactions/__init__.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023-2024 SGLang Team
+# Copyright 2025 ModelBest Inc. and/or its affiliates
+#
```
Comment (ocss884) [inline verl/interactions/__init__.py:4]: pls add the copyright of bytedance

File: verl/workers/actor/dp_actor.py (hunk 8 above)
Comment (kinza99) [reply verl/workers/actor/dp_actor.py]: In some versions, you may encounter the problem `NotImplementedError: TensorDict does not support membership checks with the `in` keyword. If you want to check if a particular key is in your TensorDict, please use key in tensordict.keys() instead.`

File: verl/interactions/base.py (hunk 9 above)
Comment (SwordFaith) [reply verl/interactions/base.py:62]: > How about `raise NotImplementedError`

The `calc_score` method is not essential for implementing interactions. Here, we use 0.0 to ensure it has no impact on the aggregated reward.

**Comment by eric-haibin-lin** [comment]
There seems to be lots of changes. I'll take some time to review as well

**Comment by SwordFaith** [comment]
Finally works, learn a lot with He Du. https://wandb.ai/swordfaith/gsm8k_async_rl/runs/8n409ugt?nw=nwuserswordfaith fin

**Comment by kinza99** [comment]
> is there a reference paper with "simulation feedback" that this PR implements?

You can refer to [https://arxiv.org/abs/2409.12917](url).

File: docs/sglang_multiturn/interaction_system.rst [hunk 12]
```diff
@@ -0,0 +1,287 @@
+Interaction System for Multi-turn RL Training
+=============================================
+
+Overview
+--------
+
+The VERL interaction system enables dynamic, multi-turn conversational feedback during reinforcement learning training. This system allows models to engage in iterative problem-solving scenarios where an interaction agent can provide corrective feedback, guidance, or evaluation based on the model's responses.
```
Comment (zhaochenyang20) [inline docs/sglang_multiturn/interaction_system.rst]: use verl inside. Do not use `VERL`.

**Comment by SwordFaith** [comment]
Before merging, please note that this PR introduces breaking changes to the multi_turn configuration. The `max_turns` parameter has been renamed to `max_assistant_turns`, retaining the same semantic meaning. Additionally, a new parameter `max_user_turns` has been added to control the number of user interactions.

File: verl/trainer/config/ppo_trainer.yaml [hunk 13]
```diff
@@ -506,11 +506,17 @@ actor_rollout_ref:
       enable: False
 
       # null for no limit (default max_length // 3)
-      max_turns: null
+      max_assistant_turns: null
```
Comment (eric-haibin-lin) [inline verl/trainer/config/ppo_trainer.yaml:509]: This looks like a breaking change, should be included in https://github.com/volcengine/verl/discussions/1902

File: verl/utils/reward_score/gsm8k.py [hunk 14]
```diff
@@ -20,12 +20,12 @@ def extract_solution(solution_str, method="strict"):
 
     if method == "strict":
         # this also tests the formatting of the model
```
Comment (eric-haibin-lin) [inline verl/utils/reward_score/gsm8k.py:22]: does this help improve the answer exaction success rate? if so we need to note that in the release note so that users are aware

File: verl/workers/megatron_workers.py [hunk 15]
```diff
@@ -465,7 +465,6 @@ def update_actor(self, data: DataProto):
         return output
 
     @register(dispatch_mode=Dispatch.DP_COMPUTE_PROTO)
```
Comment (eric-haibin-lin) [inline verl/workers/megatron_workers.py]: is this removal intended?

File: verl/trainer/ppo/ray_trainer.py [hunk 16]
```diff
@@ -460,7 +460,7 @@ def check_mutually_exclusive(mbs, mbs_per_gpu, name: str):
 
         # check multi_turn with tool config
         if config.actor_rollout_ref.rollout.multi_turn.enable:
-            assert config.actor_rollout_ref.rollout.multi_turn.tool_config_path is not None, "tool_config_path must be set when enabling multi_turn with tool, due to no role-playing support"
+            assert config.actor_rollout_ref.rollout.multi_turn.tool_config_path is not None or config.actor_rollout_ref.rollout.multi_turn.interaction_config_path is not None, "tool_config_path or interaction_config_path must be set when enabling multi_turn with tool, due to no role-playing support"
```
Comment (eric-haibin-lin) [inline verl/trainer/ppo/ray_trainer.py:478]: As the number of fields under `multi_turn` is growing, I'd suggest create a dataclass for multi_turn config similar to   https://github.com/volcengine/verl/pull/2117 and perform the assertion (validation checks) in the `__post_init__` function of the dataclass. For unit test we prefer creating dataclass directly instead of resorting to `OmegaConf.create`

File: verl/workers/rollout/schemas.py [hunk 17]
```diff
@@ -151,6 +153,16 @@ def get_generation_prompt_ids(self, tokenizer: PreTrainedTokenizer) -> list[int]
         else:
             return self.input_ids
 
+    def add_user_message(
+        self,
+        tokenizer: PreTrainedTokenizer,
+        content: str,
+    ) -> None:
```
Comment (eric-haibin-lin) [inline verl/workers/rollout/schemas.py:160]: Since this is an important class and function, i suggest adding docstring to these APIs

**Review by eric-haibin-lin — COMMENTED** [review]
I left a few comments. They do not necessarily block this PR but I hope they can be addressed in the next PR since significant progress is made in this one already

File: verl/workers/megatron_workers.py (hunk 15 above)
Comment (SwordFaith) [reply verl/workers/megatron_workers.py:467]: https://github.com/volcengine/verl/actions/runs/15785640062/job/44501336413?pr=1630 It currently triggers errors.

File: verl/trainer/ppo/ray_trainer.py (hunk 16 above)
Comment (SwordFaith) [reply verl/trainer/ppo/ray_trainer.py:478]: I will open a issue to track this https://github.com/volcengine/verl/issues/2128

File: verl/utils/reward_score/gsm8k.py (hunk 14 above)
Comment (SwordFaith) [reply verl/utils/reward_score/gsm8k.py:22]: It primarily enhances the training of small models in multi-turn user interaction scenarios, where an answer is generated for each turn. This new implementation improves the identification of the last turn by moving away from using the entire response string for reward scoring and instead focusing on message-level analysis.

File: verl/workers/rollout/schemas.py (hunk 17 above)
Comment (SwordFaith) [reply verl/workers/rollout/schemas.py:160]: I will open a issue to track this https://github.com/volcengine/verl/issues/2129.

File: verl/workers/megatron_workers.py (hunk 15 above)
Comment (SwordFaith) [reply verl/workers/megatron_workers.py]: In the previous commit, some errors were triggered. However, after the recent refactor, it seems to be working in my local environment. I have re-added the profilers.

**Review by zhaochenyang20 — APPROVED** [review]
Ready to go. Great！

**Review by tongyx361 — APPROVED** [review]

File: docs/sglang_multiturn/interaction_system.rst [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+**BaseInteraction Interface**
+
+All interaction agents must implement the ``BaseInteraction`` abstract class:
+
+.. code-block:: python
+
+    from verl.interactions.base import BaseInteraction
+    from typing import Dict, Any, List, Tuple, Optional
+
+    class BaseInteraction:
```
Comment (vermouth1992) [inline docs/sglang_multiturn/interaction_system.rst:44]: Is this just an OpenAI GymEnv?

Comment (SwordFaith) [reply docs/sglang_multiturn/interaction_system.rst:44]: I guess env or tool can both support in current tools submodule. Interaction is designed to 2 sceneraios:
- partial exposure
- inject context with workflow as user role (as many framework do)

So BaseInteraction support generate response from model api our other workflow part as endpoint. In partial exposure user requirements, typically need message history to dynamicly add user requirements. In workflow context injection, it typically need messages parsed to other node input and inject other node output back to current node. And we can use e2e outcome to train current node model (If not in current model, need to extend in finalize).

File: verl/interactions/base.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+        Returns:
+            The instance id of the tool.
+        """
+        if instance_id is None:
+            return str(uuid4())
+        else:
+            return instance_id
+
+    async def generate_response(self, instance_id: str, messages: List[Dict[str, Any]], **kwargs) -> Tuple[bool, str, float, Dict[str, Any]]:  # More clear response generation method
```
Comment (wuxibin89) [inline verl/interactions/base.py:39]: P2: use a pydantic class as return instead of Tuple

File: verl/workers/rollout/sglang_rollout/sglang_rollout.py [hunk 20]
```diff
@@ -251,6 +251,7 @@ def __init__(
             self._sgl_tools,
             self._function_call_parser,
         ) = self._initialize_tools(config, tokenizer)
+        self.interaction: dict[str, BaseInteraction] = self._intitalize_interaction(config)
```
Comment (wuxibin89) [inline verl/workers/rollout/sglang_rollout/sglang_rollout.py:254]: typo: self.interaction is not a dict

File: verl/workers/rollout/sglang_rollout/sglang_rollout.py [hunk 21]
```diff
@@ -767,10 +809,32 @@ async def _async_rollout_a_request(
...
+                            content,
+                        )
+                        if _req.interaction_kwargs and user_turns < self.config.multi_turn.max_user_turns and current_turns < self.config.multi_turn.max_assistant_turns:
+                            _req.state = AsyncRolloutRequestStateEnum.INTERACTING
+                        else:
+                            break
+            elif _req.state == AsyncRolloutRequestStateEnum.INTERACTING:
+                user_turns += 1
+                messages = [{"role": x.role, "content": x.content} for x in _req.messages]
+                should_terminate_sequence, content, reward, metrics = await self.interaction.generate_response(_req.request_id, messages, **_req.interaction_kwargs)
```
Comment (wuxibin89) [inline verl/workers/rollout/sglang_rollout/sglang_rollout.py:823]: Do we support choose different interaction based on assistant response? Or we just use one interaction, and dispatch in it?

File: verl/utils/dataset/rl_dataset.py [hunk 22]
```diff
@@ -300,11 +300,13 @@ def __getitem__(self, item):
         # add index for each prompt
         index = row_dict.get("extra_info", {}).get("index", 0)
         tools_kwargs = row_dict.get("extra_info", {}).get("tools_kwargs", {})
+        interaction_kwargs = row_dict.get("extra_info", {}).get("interaction_kwargs", {})
```
Comment (PeterSH6) [inline verl/utils/dataset/rl_dataset.py:303]: Could you add a config to control whether to include the tool related fields in dataproto? This may increase the size if tools are not used

File: verl/workers/rollout/sglang_rollout/sglang_rollout.py (hunk 21 above)
Comment (SwordFaith) [reply verl/workers/rollout/sglang_rollout/sglang_rollout.py:823]: We support different interaction cls, according to sample level config, like tools. For message history difference,should be handled inside interaction.

File: verl/interactions/base.py (hunk 19 above)
Comment (SwordFaith) [reply verl/interactions/base.py:39]: Will switch to pydantic cls in future.

File: verl/workers/rollout/sglang_rollout/sglang_rollout.py (hunk 20 above)
Comment (SwordFaith) [reply verl/workers/rollout/sglang_rollout/sglang_rollout.py:254]: It could be fixed in _initialize_interaction, which should support init multiple interactions in same config file. I guess we coulld merge 1630 first and add multiple interactions in future. Current impl already support single task training with interaction.

File: verl/utils/dataset/rl_dataset.py (hunk 22 above)
Comment (SwordFaith) [reply verl/utils/dataset/rl_dataset.py:303]: https://github.com/volcengine/verl/issues/2145 Tracked by new issue, will refactor in future.

**Comment by leoleoasd** [comment]
image
Why is user message used to calcualte the reward instead of the assistant message?

**Comment by SwordFaith** [comment]
> image Why is user message used to calcualte the reward instead of the assistant message?

We are developing a message and turn-level reward system. Calculating rewards during the user's turn doesn't necessarily mean that the user's message requires a reward. Instead, it allows the user to reallocate rewards between the assistant's turns based on the user's turn rewards.
